[ML] updating NER result to be inline with existing inference results

benwtrent opened the pull request:
This finalizes the NER inference result to bring it more inline with our other inference tasks.

For NER, the predicted_value follows the [annotated-text](https://www.elastic.co/guide/en/elasticsearch/plugins/current/mapper-annotated-text-usage.html) formatting.

Then for each entity detected, the individual prediction, probability, and position in the original sequence is returned as well.

Here are some example outputs:

```json
{
  "predicted_value" : "[Alexander](PERSON&Alexander), my name is [Benjamin Trent](PERSON&Benjamin+Trent), I work at [Acme Inc](ORG&Acme+Inc).",
  "entities" : [
    {
      "entity" : "alexander",
      "class_name" : "PERSON",
      "class_probability" : 0.9965580623850929,
      "start_pos" : 0,
      "end_pos" : 9
    },
    {
      "entity" : "benjamin trent",
      "class_name" : "PERSON",
      "class_probability" : 0.9974236537246695,
      "start_pos" : 22,
      "end_pos" : 36
    },
    {
      "entity" : "acme inc",
      "class_name" : "ORG",
      "class_probability" : 0.9985719234682839,
      "start_pos" : 48,
      "end_pos" : 56
    }
  ]
}
```

```json
{
  "predicted_value" : "this sentence has no entities",
  "entities" : [ ]
}
```

benwtrent committed:
[ML] updating NER result to be inline with existing inference results

elasticmachine commented:
Pinging @elastic/ml-core (Team:ML)

joshdevins commented:
@markharwood Could you have a look and give any feedback? This PR adds support for Annotated Text formatting on the output of our new NER task.

joshdevins commented:
Also note that Annotated Text is currently in a plugin, so actually using the field will additionally require the user to install the plugin. Are we ok with that? We can certainly document it easily.

joshdevins commented:
@benwtrent The annotations need to be encoded:

> The annotations use a markdown-like syntax using URL encoding of one or more values separated by the & symbol.

joshdevins commented:
@benwtrent I think `predicted_value_entity` should be unified with multi-class classification output, which I guess is `class_label` or something like that?

markharwood commented:
>I'm wondering if it should store the entity plus type instead? e.g. ... [Elasticsearch Inc.](Elasticsearch Inc.&organization) 

That can be [useful](https://youtu.be/_ThbDlhp1AQ?t=104) but the one thing I'd suggest changing is the formatting. If you use `Organization` instead of `organization` then it would be differentiated from any natural text e.g. this sentence:
> the key to kitchen drawer organization is in ...

Annotations are not fed through Analyzers so using a `term` query like:

```
     "term" : { "my_text_field", "Organization"}
```
..would be a way to match and highlight org annotations only - all "natural" unannotated text will have been lower-cased.
As a side-note, If I recall correctly there may be some standard for entity tagging e.g. organisations might be ORG and people PER. As long as it uses case in ways which tokenized text don't then it might be a good idea to adopt that.

benwtrent committed:
addressing PR comments

benwtrent commented:
@joshdevins 

Output is now:
```
 "[Alexander](PERSON&Alexander), my name is [Benjamin Trent](PERSON&Benjamin+Trent), I work at [Acme Inc](ORG&Acme+Inc)."
```
For the input:
```
Alexander, my name is Benjamin Trent, I work at Acme Inc.
```

Also changed the internal format to use `class_name` and `class_probability`.

I updated our `toString` for the Entity values to return the uppercased string: `MISC`, `PERSON`, `LOC`, `ORG`. This is similar to other inference result outputs I could find for NER.

joshdevins commented:
@benwtrent Since we currently only support [conll03](https://huggingface.co/datasets/conll2003) labels, can we align them?

> ner_tags: a list of classification labels, with possible values including O (0), B-PER (1), I-PER (2), B-ORG (3), I-ORG (4) B-LOC (5), I-LOC (6) B-MISC (7), I-MISC (8).

So, minus the IOB prefixes: `PER`, `ORG`, `LOC`, `MISC`

benwtrent committed:
changing PERSON to PER in default IOB NER tagging

benwtrent commented:
> When there are no entities in the input, should predicted_value be the input unaltered instead of this sentence has no entities?

@dimitris-athanasiou that WAS the input for that particular inference call :). It is indeed the input unaltered.

dimitris-athanasiou commented:
@benwtrent Hahahaha! I'll find a hole to put my head in for a while!

dimitris-athanasiou commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/NerProcessor.java:
```diff
@@ -246,9 +308,19 @@ public InferenceResults processResult(TokenizationResult tokenization, PyTorchRe
+                    )
+                );
                 startTokenIndex = endTokenIndex;
+                startFindInSeq = i;
```
Should we set this to `i + entity.length()`? Also, should we handle i being `-1` here?

dimitris-athanasiou reviewed (commented):
Looks good! Just a few things to address.

benwtrent committed:
addressing PR comments

dimitris-athanasiou reviewed (approved):
LGTM

benwtrent commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/ml-solidfy-ner-output
